src/sinks/amqp.rs

```diff
@@ -0,0 +1,615 @@
+        match self.poll_delivery_fut(cx) {
+            Poll::Pending if self.delivery_fut.len() >= SEND_RESULT_LIMIT => Poll::Pending,
+            _ => Poll::Ready(Ok(())),
+        }
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, item: Event) -> Result<(), Self::Error> {
+        assert!(
+            self.delivery_fut.len() < SEND_RESULT_LIMIT,
+            "Expected `poll_ready` to be called first."
+        );
+
+        let exchange = self.exchange.render_string(&item).map_err(|missing_keys| {
+            error!(message = "Missing keys for exchange.", missing_keys = ?missing_keys);
+        })?;
+
+        let (key, body) = encode_event(item, &self.key_field, &self.encoding);
+
+        let seqno = self.seq_head;
+        self.seq_head += 1;
+
+        let channel = self.channel.clone();
+        self.delivery_fut.push(Box::pin(async move {
+            let f = channel
+                .basic_publish(
+                    &exchange,
+                    &key,
+                    BasicPublishOptions::default(),
+                    body,
+                    BasicProperties::default(),
+                );
+
+            (seqno, f)
+        }));
+
+        Ok(())
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        ready!(self.poll_delivery_fut(cx));
+
+        let this = Pin::into_inner(self);
+        while !this.in_flight.is_empty() {
+            match ready!(Pin::new(&mut this.in_flight).poll_next(cx)) {
+                Some((seqno, Ok(result))) => {
+                    match result {
+                        Ok(opt_msg) => {
+                            if let Some(msg) = opt_msg {
+                                error!(message = "Message was not acknowledged", msg.delivery_tag)
+                            }
+                        }
+                        Err(error) => error!(message = "Amqp error.", %error),
+                    };
+
+                    this.pending_acks.insert(seqno);
+
+                    let mut num_to_ack = 0;
+                    while this.pending_acks.remove(&this.seq_tail) {
+                        num_to_ack += 1;
+                        this.seq_tail += 1
+                    }
+                    this.acker.ack(num_to_ack);
+                }
+                Some((_, Err(Canceled))) => {
+                    error!(message = "Request canceled.");
+                    return Poll::Ready(Err(()));
+                }
+                None => break,
+            }
+        }
+
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.poll_flush(cx)
+    }
+}
+
+async fn healthcheck(_config: AmqpSinkConfig, channel: Arc<lapin::Channel>) -> crate::Result<()> {
+    trace!("Healthcheck started.");
+
+    if !channel.status().connected() {
+        warn!("Channel is not connected");
```

Comment:
I think we can drop this `warn`. Returning the error should be enough to get useful output in the vector logs.